### ISSUE TITLE
Questions about CSE Loss

### ISSUE THREAD

**Issue by fmy7834** [richard-peng-xia/LMPT#7]
Hello, I 'm interested in your excellent work. And I have two questions when re-produce your work.
1. In line 210 of lmpt/train.py: 
   **class_weights = sfm(torch.from_numpy(np.asarray(mmcv.load(freq_file)['class_freq'])).to(torch.float32).cuda())**,
    Why apply softmax on class_counts? After applying it, the result contains zeros, which influence the calculation of margin and 
    class_weights in csel.py (class_counts is as denominator). For example, for voc-lt, class_counts after softmax is like this:  
    ![image](https://github.com/richard-peng-xia/LMPT/assets/80671862/e97a346c-b6e4-4ff7-8262-157ee845f257)

2. In line 20 of lmpt/csel.py:
    **self.class_counts = self.class_counts.squeeze(dim=0).expand(labels.shape[0], self.class_counts.shape[-1])**,
    after this line, self.class_counts has two dimensions: [batch_size, n_cls]. But in line 25, it requires self.class_counts to have a 
    third dimension in torch.sum(dim=2):
    **class_weights = (1 / self.class_counts) ** self.gamma / torch.sum((1 / self.class_counts) ** self.gamma, dim=2)**
    I think the line 25 should be like this:
    **class_weights = (1 / self.class_counts) ** self.gamma / torch.sum((1 / self.class_counts) ** self.gamma, dim=1, 
     keepdim=True)**

I would appreciate it if you could answer my confusions.

**Comment by richard-peng-xia** [comment]
Hi @fmy7834 ,

Sorry for the late response.

1. As I mentioned in issue #3, using softmax for multi-label tasks is not optimal. You can improve performance by using functions such as sigmoid.

2. The code in this step is fine. For specific tensor shapes, please refer to issue #2, where I have provided detailed information on the shapes.

Thank you for your attention.
